# Assignment 1  
Simon Karumbi  
s3455453  
### Question 1
Import and preprocess Credit Approval Data from [here](https://archive.ics.uci.edu/ml/machine-learning-databases/credit-screening)

### Attribute Information
A1: b, a

A2: continuous

A3: continuous

A4: u, y, l, t

A5: g, p, gg
A6: c, d, cc, i, j, k, m, r, q, w, x, e, aa, ff

A7: v, h, bb, j, n, z, dd, ff, o

A8: continuous

A9: t, f

A10: t, f

A11: continuous

A12: t, f

A13: g, p, s

A14: continuous

A15: continuous

A16: +,- (class attribute)

In [1]:
import pandas as pd
import numpy as np
import os, ssl

if (not os.environ.get('PYTHONHTTPSVERIFY', '') and
    getattr(ssl, '_create_unverified_context', None)): 
    ssl._create_default_https_context = ssl._create_unverified_context

In [2]:
attributeNames = ['A' + str(i+1) for i in range(16)]

In [3]:
data = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/credit-screening/crx.data',
                   names = attributeNames,
                   header = None,
                   na_values = ['?'])

In [4]:
data.shape

(690, 16)

In [5]:
data.head(15)

A1     A2      A3 A4 A5  A6 A7     A8 A9 A10  A11 A12 A13    A14    A15 A16
0   b  30.83   0.000  u  g   w  v  1.250  t   t    1   f   g  202.0      0   +
1   a  58.67   4.460  u  g   q  h  3.040  t   t    6   f   g   43.0    560   +
2   a  24.50   0.500  u  g   q  h  1.500  t   f    0   f   g  280.0    824   +
3   b  27.83   1.540  u  g   w  v  3.750  t   t    5   t   g  100.0      3   +
4   b  20.17   5.625  u  g   w  v  1.710  t   f    0   f   s  120.0      0   +
5   b  32.08   4.000  u  g   m  v  2.500  t   f    0   t   g  360.0      0   +
6   b  33.17   1.040  u  g   r  h  6.500  t   f    0   t   g  164.0  31285   +
7   a  22.92  11.585  u  g  cc  v  0.040  t   f    0   f   g   80.0   1349   +
8   b  54.42   0.500  y  p   k  h  3.960  t   f    0   f   g  180.0    314   +
9   b  42.50   4.915  y  p   w  v  3.165  t   f    0   t   g   52.0   1442   +
10  b  22.08   0.830  u  g   c  h  2.165  f   f    0   t   g  128.0      0   +
11  b  29.92   1.835  u  g   c  h  4.335  t   f    0   f   g  260.0    200   +
12  a  38.25   6.000  u  g   k  v  1.000  t   f    0   t   g    0.0      0   +
13  b  48.08   6.040  u  g   k  v  0.040  f   f    0   f   g    0.0   2690   +
14  a  45.83  10.500  u  g   q  v  5.000  t   t    7   t   g    0.0      0   +

In [6]:
data.dtypes

A1      object
A2     float64
A3     float64
A4      object
A5      object
A6      object
A7      object
A8     float64
A9      object
A10     object
A11      int64
A12     object
A13     object
A14    float64
A15      int64
A16     object
dtype: object

In [7]:
data.describe(include = np.number).round(3)

A2       A3       A8      A11       A14         A15
count  678.000  690.000  690.000  690.000   677.000     690.000
mean    31.568    4.759    2.223    2.400   184.015    1017.386
std     11.958    4.978    3.347    4.863   173.807    5210.103
min     13.750    0.000    0.000    0.000     0.000       0.000
25%     22.602    1.000    0.165    0.000    75.000       0.000
50%     28.460    2.750    1.000    0.000   160.000       5.000
75%     38.230    7.208    2.625    3.000   276.000     395.500
max     80.250   28.000   28.500   67.000  2000.000  100000.000

In [8]:
data.describe(include = np.object)

A1   A4   A5   A6   A7   A9  A10  A12  A13  A16
count   678  684  684  681  681  690  690  690  690  690
unique    2    3    3   14    9    2    2    2    3    2
top       b    u    g    c    v    t    f    f    g    -
freq    468  519  519  137  399  361  395  374  625  383

Observing the categorical data types

In [9]:
for col in data.select_dtypes(include = 'object'):
    print('Category ' + col + ' : ')
    print(data[col].value_counts(),'\n')

Category A1 : 
b    468
a    210
Name: A1, dtype: int64 

Category A4 : 
u    519
y    163
l      2
Name: A4, dtype: int64 

Category A5 : 
g     519
p     163
gg      2
Name: A5, dtype: int64 

Category A6 : 
c     137
q      78
w      64
i      59
aa     54
ff     53
k      51
cc     41
x      38
m      38
d      30
e      25
j      10
r       3
Name: A6, dtype: int64 

Category A7 : 
v     399
h     138
bb     59
ff     57
z       8
j       8
dd      6
n       4
o       2
Name: A7, dtype: int64 

Category A9 : 
t    361
f    329
Name: A9, dtype: int64 

Category A10 : 
f    395
t    295
Name: A10, dtype: int64 

Category A12 : 
f    374
t    316
Name: A12, dtype: int64 

Category A13 : 
g    625
s     57
p      8
Name: A13, dtype: int64 

Category A16 : 
-    383
+    307
Name: A16, dtype: int64 



### Eliminating Redundant Features
The dataset does not contain any
- ID like features
- Constant features
- Date or Time features

However, columns A4 and A5 seem to have the same amount of variables, with the exact same value count distribution. This indicates there might be a repeated column.

In [10]:
A4 = data['A4'].copy()
A5 = data['A5'].copy()

A5 = A5.replace({'g':'u', 'p':'y', 'gg':'l'})

print(A4.head(10).values)
print(A5.head(10).values)

print(A4.equals(A5))

['u' 'u' 'u' 'u' 'u' 'u' 'u' 'u' 'y' 'y']
['u' 'u' 'u' 'u' 'u' 'u' 'u' 'u' 'y' 'y']
True


We can see that the columns A4 and A5 are the same, therefore, we can drop either column as to prevent redundancy.

In [11]:
data.drop(columns = 'A5', inplace=True)

In [12]:
for col in (data.select_dtypes(include = 'object')):
    print(data[col].unique())

['b' 'a' nan]
['u' 'y' nan 'l']
['w' 'q' 'm' 'r' 'cc' 'k' 'c' 'd' 'x' 'i' 'e' 'aa' 'ff' 'j' nan]
['v' 'h' 'bb' 'ff' 'j' 'z' nan 'o' 'dd' 'n']
['t' 'f']
['t' 'f']
['f' 't']
['g' 's' 'p']
['+' '-']


### Imputing Values
Lets impute integers with their __median__ and categorical variables with their __mode__

In [13]:
data_imputed = data.copy()

for col in data_imputed.select_dtypes(include = 'object'):
    mode = data_imputed[col].mode()[0]
    print('Category ' + col + ' mode: ' + mode)
    data_imputed[col] =  data_imputed[col].fillna(mode)

Category A1 mode: b
Category A4 mode: u
Category A6 mode: c
Category A7 mode: v
Category A9 mode: t
Category A10 mode: f
Category A12 mode: f
Category A13 mode: g
Category A16 mode: -


In [14]:
for col in data_imputed.select_dtypes(include = 'number'):
    median = np.round(data_imputed[col].median(),3)
    print('Category ' + col + ' median: ' + str(median))
    data_imputed[col] =  data_imputed[col].fillna(median)

Category A2 median: 28.46
Category A3 median: 2.75
Category A8 median: 1.0
Category A11 median: 0.0
Category A14 median: 160.0
Category A15 median: 5.0


In [15]:
data_imputed.isna().sum()

A1     0
A2     0
A3     0
A4     0
A6     0
A7     0
A8     0
A9     0
A10    0
A11    0
A12    0
A13    0
A14    0
A15    0
A16    0
dtype: int64

All NA values have now been removed from the dataset

### Integer Encoding
Let's discretise column A2 by equal-frequency binning with "low", "medium", and "high" and then apply integer encoding.

We can also go ahead and perform a manual replace encoding for the A16 target feature, where we assume that '+' indicates a credit approval (1), and '-' indicates no approval (0).

In [16]:
data_encoded = data_imputed.copy()
data_encoded['A2'] = pd.qcut(data_encoded['A2'], q = 3, labels =['low','medium','high'])

In [17]:
data_encoded['A2'].value_counts()

medium    231
low       230
high      229
Name: A2, dtype: int64

In [18]:
level_mapping = {'low':0, 'medium':1, 'high':2 }
data_encoded['A2'] = data_encoded['A2'].replace(level_mapping)
data_encoded['A16'] = data_encoded['A16'].replace({'+':1, '-':0})

In [19]:
data_encoded.head()

A1  A2     A3 A4 A6 A7    A8 A9 A10  A11 A12 A13    A14  A15  A16
0  b   1  0.000  u  w  v  1.25  t   t    1   f   g  202.0    0    1
1  a   2  4.460  u  q  h  3.04  t   t    6   f   g   43.0  560    1
2  a   1  0.500  u  q  h  1.50  t   f    0   f   g  280.0  824    1
3  b   1  1.540  u  w  v  3.75  t   t    5   t   g  100.0    3    1
4  b   0  5.625  u  w  v  1.71  t   f    0   f   s  120.0    0    1

### One Hot Encoding
Now Category A2 has been integer encoded, we can go ahead and perform one hot encoding on the rest of the categorical features. We can separate the target feature, Category A16 from the rest of the dataset.

In [20]:
# Removal of the target feature, to be merged later
target = data_encoded['A16']
data_ohe = data_encoded.drop(columns = 'A16')

In [21]:
for col in data_ohe.select_dtypes(include ='object'):
    if(len(data_ohe[col].unique()) ==  2):
        data_ohe[col] = pd.get_dummies(data_ohe[col], drop_first = True)

data_ohe = pd.get_dummies(data_ohe)     

print(data_ohe.shape)

(690, 39)


In [22]:
data_ohe.head()

A1  A2     A3    A8  A9  A10  A11  A12    A14  A15  ...  A7_ff  A7_h  A7_j  \
0   1   1  0.000  1.25   1    1    1    0  202.0    0  ...      0     0     0   
1   0   2  4.460  3.04   1    1    6    0   43.0  560  ...      0     1     0   
2   0   1  0.500  1.50   1    0    0    0  280.0  824  ...      0     1     0   
3   1   1  1.540  3.75   1    1    5    1  100.0    3  ...      0     0     0   
4   1   0  5.625  1.71   1    0    0    0  120.0    0  ...      0     0     0   

   A7_n  A7_o  A7_v  A7_z  A13_g  A13_p  A13_s  
0     0     0     1     0      1      0      0  
1     0     0     0     0      1      0      0  
2     0     0     0     0      1      0      0  
3     0     0     1     0      1      0      0  
4     0     0     1     0      0      0      1  

[5 rows x 39 columns]

### Scaling
Scale the features for use in SciKitLearn and export to CSV

In [23]:
from sklearn import preprocessing

data_scaled = data_ohe.copy()

# Saving the column names prior to scaling
col_names = data_scaled.columns

# Perform standardised scaling
data_standard_scaled = preprocessing.StandardScaler().fit_transform(data_scaled)

# Convert back to a Data Frame, with correct column names
data_scaled = pd.DataFrame(data_standard_scaled,
                    columns = col_names)

print(data_scaled.shape)

data_scaled.sample(n=15, random_state = 1)

(690, 39)


A1        A2        A3        A8        A9       A10       A11  \
666 -1.511858 -1.224303  1.405407 -0.590118 -1.047504 -0.864196 -0.493887   
520  0.661438 -1.224303  0.551059 -0.216324  0.954650  1.157144 -0.288101   
681  0.661438  0.001777 -0.755590  0.232229 -1.047504 -0.864196 -0.493887   
23  -1.511858  0.001777  1.958221  0.257647  0.954650  1.157144 -0.288101   
65   0.661438  1.227857 -0.647038 -0.204362  0.954650  1.157144 -0.288101   
339  0.661438  0.001777 -0.353544 -0.440600 -1.047504 -0.864196 -0.493887   
56   0.661438 -1.224303 -0.956613 -0.639459  0.954650 -0.864196 -0.493887   
394  0.661438  1.227857 -0.705334 -0.590118 -1.047504 -0.864196 -0.493887   
47   0.661438  0.001777 -0.060051  1.141296  0.954650  1.157144  0.123472   
358  0.661438  0.001777 -0.353544 -0.615536 -1.047504 -0.864196 -0.493887   
442  0.661438  0.001777 -0.411841 -0.627497 -1.047504 -0.864196 -0.493887   
299  0.661438 -1.224303  1.480791  0.332406 -1.047504  1.157144 -0.082314   
414  0.661438 -1.224303 -0.864143 -0.627497 -1.047504 -0.864196 -0.493887   
274  0.661438  0.001777 -0.454056  0.007953 -1.047504 -0.864196 -0.493887   
676 -1.511858  0.001777  1.187297 -0.639459 -1.047504  1.157144  1.975546   

          A12       A14       A15  ...     A7_ff  A7_h      A7_j     A7_n  \
666  1.087908 -0.020703 -0.195413  ... -0.300079  -0.5 -0.108306 -0.07636   
520 -0.919195 -0.136938 -0.150468  ... -0.300079  -0.5 -0.108306 -0.07636   
681 -0.919195 -0.043950 -0.092270  ... -0.300079   2.0 -0.108306 -0.07636   
23  -0.919195 -0.369408 -0.193301  ... -0.300079   2.0 -0.108306 -0.07636   
65   1.087908  1.955289  9.408288  ... -0.300079  -0.5 -0.108306 -0.07636   
339  1.087908  0.676706 -0.182544  ... -0.300079  -0.5 -0.108306 -0.07636   
56  -0.919195 -1.066817 -0.195413  ... -0.300079  -0.5 -0.108306 -0.07636   
394 -0.919195 -1.066817 -0.157959  ... -0.300079  -0.5 -0.108306 -0.07636   
47  -0.919195  0.740635 -0.137791  ... -0.300079   2.0 -0.108306 -0.07636   
358  1.087908 -0.369408 -0.195413  ... -0.300079  -0.5 -0.108306 -0.07636   
442  1.087908 -0.601877 -0.195413  ... -0.300079  -0.5 -0.108306 -0.07636   
299  1.087908 -0.020703 -0.162185  ... -0.300079  -0.5 -0.108306 -0.07636   
414  1.087908  0.560471 -0.184849  ... -0.300079  -0.5 -0.108306 -0.07636   
274  1.087908  0.909175 -0.195413  ... -0.300079   2.0 -0.108306 -0.07636   
676  1.087908 -0.317102 -0.194837  ... -0.300079   2.0 -0.108306 -0.07636   

         A7_o      A7_v      A7_z     A13_g     A13_p     A13_s  
666 -0.053916  0.831370 -0.108306  0.322490 -0.108306 -0.300079  
520 -0.053916  0.831370 -0.108306  0.322490 -0.108306 -0.300079  
681 -0.053916 -1.202834 -0.108306  0.322490 -0.108306 -0.300079  
23  -0.053916 -1.202834 -0.108306  0.322490 -0.108306 -0.300079  
65  -0.053916  0.831370 -0.108306  0.322490 -0.108306 -0.300079  
339 -0.053916  0.831370 -0.108306  0.322490 -0.108306 -0.300079  
56  -0.053916  0.831370 -0.108306  0.322490 -0.108306 -0.300079  
394 -0.053916  0.831370 -0.108306  0.322490 -0.108306 -0.300079  
47  -0.053916 -1.202834 -0.108306  0.322490 -0.108306 -0.300079  
358 -0.053916  0.831370 -0.108306  0.322490 -0.108306 -0.300079  
442 -0.053916  0.831370 -0.108306 -3.100868 -0.108306  3.332456  
299 -0.053916  0.831370 -0.108306  0.322490 -0.108306 -0.300079  
414 -0.053916  0.831370 -0.108306  0.322490 -0.108306 -0.300079  
274 -0.053916 -1.202834 -0.108306 -3.100868 -0.108306  3.332456  
676 -0.053916 -1.202834 -0.108306  0.322490 -0.108306 -0.300079  

[15 rows x 39 columns]

### Re-adding Target Feature
Let's add the target feature back to the standardised data set. The final dataset has been fully cleaned.

In [24]:
df_clean = data_scaled.assign(target = target.values)

df_clean = df_clean.round(3)

df_clean.sample(n = 15, random_state = 1)

A1     A2     A3     A8     A9    A10    A11    A12    A14    A15  \
666 -1.512 -1.224  1.405 -0.590 -1.048 -0.864 -0.494  1.088 -0.021 -0.195   
520  0.661 -1.224  0.551 -0.216  0.955  1.157 -0.288 -0.919 -0.137 -0.150   
681  0.661  0.002 -0.756  0.232 -1.048 -0.864 -0.494 -0.919 -0.044 -0.092   
23  -1.512  0.002  1.958  0.258  0.955  1.157 -0.288 -0.919 -0.369 -0.193   
65   0.661  1.228 -0.647 -0.204  0.955  1.157 -0.288  1.088  1.955  9.408   
339  0.661  0.002 -0.354 -0.441 -1.048 -0.864 -0.494  1.088  0.677 -0.183   
56   0.661 -1.224 -0.957 -0.639  0.955 -0.864 -0.494 -0.919 -1.067 -0.195   
394  0.661  1.228 -0.705 -0.590 -1.048 -0.864 -0.494 -0.919 -1.067 -0.158   
47   0.661  0.002 -0.060  1.141  0.955  1.157  0.123 -0.919  0.741 -0.138   
358  0.661  0.002 -0.354 -0.616 -1.048 -0.864 -0.494  1.088 -0.369 -0.195   
442  0.661  0.002 -0.412 -0.627 -1.048 -0.864 -0.494  1.088 -0.602 -0.195   
299  0.661 -1.224  1.481  0.332 -1.048  1.157 -0.082  1.088 -0.021 -0.162   
414  0.661 -1.224 -0.864 -0.627 -1.048 -0.864 -0.494  1.088  0.560 -0.185   
274  0.661  0.002 -0.454  0.008 -1.048 -0.864 -0.494  1.088  0.909 -0.195   
676 -1.512  0.002  1.187 -0.639 -1.048  1.157  1.976  1.088 -0.317 -0.195   

     ...  A7_h   A7_j   A7_n   A7_o   A7_v   A7_z  A13_g  A13_p  A13_s  target  
666  ...  -0.5 -0.108 -0.076 -0.054  0.831 -0.108  0.322 -0.108 -0.300       0  
520  ...  -0.5 -0.108 -0.076 -0.054  0.831 -0.108  0.322 -0.108 -0.300       1  
681  ...   2.0 -0.108 -0.076 -0.054 -1.203 -0.108  0.322 -0.108 -0.300       0  
23   ...   2.0 -0.108 -0.076 -0.054 -1.203 -0.108  0.322 -0.108 -0.300       1  
65   ...  -0.5 -0.108 -0.076 -0.054  0.831 -0.108  0.322 -0.108 -0.300       1  
339  ...  -0.5 -0.108 -0.076 -0.054  0.831 -0.108  0.322 -0.108 -0.300       0  
56   ...  -0.5 -0.108 -0.076 -0.054  0.831 -0.108  0.322 -0.108 -0.300       1  
394  ...  -0.5 -0.108 -0.076 -0.054  0.831 -0.108  0.322 -0.108 -0.300       0  
47   ...   2.0 -0.108 -0.076 -0.054 -1.203 -0.108  0.322 -0.108 -0.300       1  
358  ...  -0.5 -0.108 -0.076 -0.054  0.831 -0.108  0.322 -0.108 -0.300       0  
442  ...  -0.5 -0.108 -0.076 -0.054  0.831 -0.108 -3.101 -0.108  3.332       0  
299  ...  -0.5 -0.108 -0.076 -0.054  0.831 -0.108  0.322 -0.108 -0.300       0  
414  ...  -0.5 -0.108 -0.076 -0.054  0.831 -0.108  0.322 -0.108 -0.300       0  
274  ...   2.0 -0.108 -0.076 -0.054 -1.203 -0.108 -3.101 -0.108  3.332       0  
676  ...   2.0 -0.108 -0.076 -0.054 -1.203 -0.108  0.322 -0.108 -0.300       0  

[15 rows x 40 columns]

__The dataset is now clean.__ It has no missing values, all columns are numerical and have been scaled, the target feature is correctly label encoded. It can now be used in a SciKitLearn model without any further data preprocessing.

### Summary and Export

In [25]:
df_clean.shape

(690, 40)

In [26]:
df_clean.describe(include='all').round(3) 

A1       A2       A3       A8       A9      A10      A11      A12  \
count  690.000  690.000  690.000  690.000  690.000  690.000  690.000  690.000   
mean    -0.000    0.000    0.000    0.000   -0.000    0.000   -0.000    0.000   
std      1.001    1.001    1.001    1.001    1.001    1.001    1.001    1.001   
min     -1.512   -1.224   -0.957   -0.665   -1.048   -0.864   -0.494   -0.919   
25%     -1.512   -1.224   -0.756   -0.616   -1.048   -0.864   -0.494   -0.919   
50%      0.661    0.002   -0.404   -0.366    0.955   -0.864   -0.494   -0.919   
75%      0.661    1.228    0.493    0.120    0.955    1.157    0.123    1.088   
max      0.661    1.228    4.672    7.858    0.955    1.157   13.294    1.088   

           A14      A15  ...     A7_h     A7_j     A7_n     A7_o     A7_v  \
count  690.000  690.000  ...  690.000  690.000  690.000  690.000  690.000   
mean    -0.000    0.000  ...    0.000    0.000    0.000   -0.000   -0.000   
std      1.001    1.001  ...    1.001    1.001    1.001    1.001    1.001   
min     -1.067   -0.195  ...   -0.500   -0.108   -0.076   -0.054   -1.203   
25%     -0.602   -0.195  ...   -0.500   -0.108   -0.076   -0.054   -1.203   
50%     -0.137   -0.194  ...   -0.500   -0.108   -0.076   -0.054    0.831   
75%      0.514   -0.119  ...   -0.500   -0.108   -0.076   -0.054    0.831   
max     10.557   19.012  ...    2.000    9.233   13.096   18.547    0.831   

          A7_z    A13_g    A13_p    A13_s   target  
count  690.000  690.000  690.000  690.000  690.000  
mean     0.000   -0.000    0.000    0.000    0.445  
std      1.001    1.001    1.001    1.001    0.497  
min     -0.108   -3.101   -0.108   -0.300    0.000  
25%     -0.108    0.322   -0.108   -0.300    0.000  
50%     -0.108    0.322   -0.108   -0.300    0.000  
75%     -0.108    0.322   -0.108   -0.300    1.000  
max      9.233    0.322    9.233    3.332    1.000  

[8 rows x 40 columns]

In [27]:
df_clean.head(5)

A1     A2     A3     A8     A9    A10    A11    A12    A14    A15  ...  \
0  0.661  0.002 -0.957 -0.291  0.955  1.157 -0.288 -0.919  0.107 -0.195  ...   
1 -1.512  1.228 -0.060  0.244  0.955  1.157  0.741 -0.919 -0.817 -0.088  ...   
2 -1.512  0.002 -0.856 -0.216  0.955 -0.864 -0.494 -0.919  0.560 -0.037  ...   
3  0.661  0.002 -0.647  0.457  0.955  1.157  0.535  1.088 -0.486 -0.195  ...   
4  0.661 -1.224  0.174 -0.154  0.955 -0.864 -0.494 -0.919 -0.369 -0.195  ...   

   A7_h   A7_j   A7_n   A7_o   A7_v   A7_z  A13_g  A13_p  A13_s  target  
0  -0.5 -0.108 -0.076 -0.054  0.831 -0.108  0.322 -0.108 -0.300       1  
1   2.0 -0.108 -0.076 -0.054 -1.203 -0.108  0.322 -0.108 -0.300       1  
2   2.0 -0.108 -0.076 -0.054 -1.203 -0.108  0.322 -0.108 -0.300       1  
3  -0.5 -0.108 -0.076 -0.054  0.831 -0.108  0.322 -0.108 -0.300       1  
4  -0.5 -0.108 -0.076 -0.054  0.831 -0.108 -3.101 -0.108  3.332       1  

[5 rows x 40 columns]

In [28]:
df_clean.to_csv('df_clean.csv', index = False)

# Question 2


### Exercise 5
Setting up the data frame 

In [29]:
country = [
    'Afghanistan',
    'Haiti',
    'Nigeria',
    'Egypt',
    'Argentina',
    'China',
    'Brazil',
    'Israel',
    'USA',
    'Ireland',
    'UK',
    'Germany',
    'Canada',
    'Australia',
    'Sweden',
    'New Zealand'
]

In [30]:
Afghanistan = [59.61,23.21,74.30,4.44,0.40,1.5171]
Haiti = [45.00,47.67,73.10,0.09,3.40,1.7999]
Nigeria = [51.30,38.23,82.60,1.07,4.10,2.4493]
Egypt = [70.48,26.58,19.60,1.86,5.30,2.8622]
Argentina = [75.77,32.30,13.30,0.76,10.10,2.9961]
China = [74.87,29.98,13.70,1.95,6.40,3.6356]
Brazil = [73.12,42.93,14.50,1.43,7.20,3.7741]
Israel = [81.30,28.80,3.60,6.77,12.50,5.8069]
USA = [78.51,29.85,6.30,4.72,13.70,7.1357]
Ireland = [80.15,27.23,3.50,0.60,11.50,7.5360]
UK = [80.09,28.49,4.40,2.59,13.00,7.7751]
Germany = [80.24,22.07,3.50,1.31,12.00,8.0461]
Canada = [80.99,24.79,4.90,1.42,14.20,8.6725]
Australia = [82.09,25.40,4.20,1.86,11.50,8.8442]
Sweden = [81.43,22.18,2.40,1.27,12.80,9.2985]
NewZealand = [80.67,27.81,4.90,1.13,12.30,9.4627]

In [31]:
col_names = [
    'life_exp',
    'top10_income',
    'infant_mort',
    'mil_spend',
    'school_years',
    'cpi'
]

In [32]:
cpi = pd.DataFrame([Afghanistan,Haiti,Nigeria,Egypt,Argentina,China,Brazil,Israel,USA,Ireland,UK,Germany,Canada,Australia,Sweden,NewZealand], columns = col_names, index = country)

In [33]:
cpi

life_exp  top10_income  infant_mort  mil_spend  school_years  \
Afghanistan     59.61         23.21         74.3       4.44           0.4   
Haiti           45.00         47.67         73.1       0.09           3.4   
Nigeria         51.30         38.23         82.6       1.07           4.1   
Egypt           70.48         26.58         19.6       1.86           5.3   
Argentina       75.77         32.30         13.3       0.76          10.1   
China           74.87         29.98         13.7       1.95           6.4   
Brazil          73.12         42.93         14.5       1.43           7.2   
Israel          81.30         28.80          3.6       6.77          12.5   
USA             78.51         29.85          6.3       4.72          13.7   
Ireland         80.15         27.23          3.5       0.60          11.5   
UK              80.09         28.49          4.4       2.59          13.0   
Germany         80.24         22.07          3.5       1.31          12.0   
Canada          80.99         24.79          4.9       1.42          14.2   
Australia       82.09         25.40          4.2       1.86          11.5   
Sweden          81.43         22.18          2.4       1.27          12.8   
New Zealand     80.67         27.81          4.9       1.13          12.3   

                cpi  
Afghanistan  1.5171  
Haiti        1.7999  
Nigeria      2.4493  
Egypt        2.8622  
Argentina    2.9961  
China        3.6356  
Brazil       3.7741  
Israel       5.8069  
USA          7.1357  
Ireland      7.5360  
UK           7.7751  
Germany      8.0461  
Canada       8.6725  
Australia    8.8442  
Sweden       9.2985  
New Zealand  9.4627

In [45]:
Russia = (67.62, 31.68, 10.00, 3.87, 12.90, np.nan)

### Calculating the Manhattan Distance
Question 3a. Calculating the Manhattan Distance of for the three nearest neighbours of Russia

In [35]:
# Defining the function to caclulate 3NN
def manhattan(row, instance):
    difference = 0
    row, instance = np.array(row), np.array(instance)
    
    for i in range(len(row) - 1):
        difference += abs(row[i] - instance[i])
        
    return difference

In [36]:
# Creating an array to hold the differences between each of the countries in the dataframe
difference = []
for i in range(len(cpi)):
    difference.append(manhattan(cpi.iloc[i][:-1],Russia[:-1]))

In [37]:
# Adding differences as a new column and sorting from smallest to largest Manhattan Distance
cpi['difference'] = difference
cpi = cpi.sort_values(by=['difference'])
cpi

life_exp  top10_income  infant_mort  mil_spend  school_years  \
China           74.87         29.98         13.7       1.95           6.4   
Argentina       75.77         32.30         13.3       0.76          10.1   
USA             78.51         29.85          6.3       4.72          13.7   
Egypt           70.48         26.58         19.6       1.86           5.3   
UK              80.09         28.49          4.4       2.59          13.0   
Brazil          73.12         42.93         14.5       1.43           7.2   
New Zealand     80.67         27.81          4.9       1.13          12.3   
Israel          81.30         28.80          3.6       6.77          12.5   
Ireland         80.15         27.23          3.5       0.60          11.5   
Canada          80.99         24.79          4.9       1.42          14.2   
Australia       82.09         25.40          4.2       1.86          11.5   
Germany         80.24         22.07          3.5       1.31          12.0   
Sweden          81.43         22.18          2.4       1.27          12.8   
Afghanistan     59.61         23.21         74.3       4.44           0.4   
Nigeria         51.30         38.23         82.6       1.07           4.1   
Haiti           45.00         47.67         73.1       0.09           3.4   

                cpi  difference  
China        3.6356       14.57  
Argentina    2.9961       15.18  
USA          7.1357       17.27  
Egypt        2.8622       19.57  
UK           7.7751       22.54  
Brazil       3.7741       23.69  
New Zealand  9.4627       24.76  
Israel       5.8069       25.86  
Ireland      7.5360       26.75  
Canada       8.6725       27.81  
Australia    8.8442       28.56  
Germany      8.0461       31.29  
Sweden       9.2985       33.51  
Afghanistan  1.5171       81.35  
Nigeria      2.4493       98.27  
Haiti        1.7999      105.49

In [38]:
# 3 nearest neighbours
knn = 3
estimate = 0

for i in range(knn):
    estimate += cpi.iloc[i].cpi
estimate = estimate/knn

print(estimate)

4.589133333333334


__Answer: The estimate for Russia's CPI using 3 Nearest Neighbours is 4.5891__

### Calculating Weighted KNN 
Question 3b. Calculating the __weighted knn__ where k = 16 

In [39]:
knn = 16 
estimate = 0 

for i in range(knn):
    estimate += cpi.iloc[i].cpi/(i+1)

estimate = estimate/knn
print(estimate)

1.0480002437926657


__Answer: The CPI for Russia when using weighted KNN over the full data set is 1.0480__

### Calcluating a Standardised, 3 Nearest Neighbours
Question 3b. Calculating the 3 nearest neighbours when using Range Normalisation to Standardise the Dataset

In [40]:
# Creating a copy of the dataset for Normalisation
cpi_copy = cpi.copy()

In [41]:
# Range Normalisation
for column in cpi.columns:
    max_value = cpi[column].max()
    min_value = cpi[column].min()
    cpi_copy[column] = (cpi[column] - min_value) / (max_value - min_value)
cpi_copy

life_exp  top10_income  infant_mort  mil_spend  school_years  \
China        0.805338      0.308984     0.140898   0.278443      0.434783   
Argentina    0.829604      0.399609     0.135910   0.100299      0.702899   
USA          0.903478      0.303906     0.048628   0.693114      0.963768   
Egypt        0.686978      0.176172     0.214464   0.264970      0.355072   
UK           0.946077      0.250781     0.024938   0.374251      0.913043   
Brazil       0.758156      0.814844     0.150873   0.200599      0.492754   
New Zealand  0.961715      0.224219     0.031172   0.155689      0.862319   
Israel       0.978700      0.262891     0.014963   1.000000      0.876812   
Ireland      0.947695      0.201563     0.013716   0.076347      0.804348   
Canada       0.970342      0.106250     0.031172   0.199102      1.000000   
Australia    1.000000      0.130078     0.022444   0.264970      0.804348   
Germany      0.950121      0.000000     0.013716   0.182635      0.840580   
Sweden       0.982205      0.004297     0.000000   0.176647      0.898551   
Afghanistan  0.393907      0.044531     0.896509   0.651198      0.000000   
Nigeria      0.169857      0.631250     1.000000   0.146707      0.268116   
Haiti        0.000000      1.000000     0.881546   0.000000      0.217391   

                  cpi  difference  
China        0.266626    0.000000  
Argentina    0.186141    0.006709  
USA          0.707134    0.029696  
Egypt        0.169289    0.054993  
UK           0.787606    0.087659  
Brazil       0.284057    0.100308  
New Zealand  1.000000    0.112077  
Israel       0.539896    0.124175  
Ireland      0.757514    0.133964  
Canada       0.900549    0.145623  
Australia    0.922158    0.153872  
Germany      0.821713    0.183898  
Sweden       0.979334    0.208315  
Afghanistan  0.000000    0.734492  
Nigeria      0.117323    0.920590  
Haiti        0.035592    1.000000

In [42]:
# 3 nearest neighbours
knn = 3
estimate = 0

for i in range(knn):
    estimate += cpi_copy.iloc[i].cpi
estimate = estimate/knn

estimate = estimate * cpi['cpi'].max()
print(estimate)

3.6585946716841184


__Answer: 3.6586 is the estimated CPI of Russie using 3 nearest neighbours, standardised through range normalisation__

### Calculating the weighted Nearest Neighbour for Russia using Range Normalisation
Question 3d. What would the CPI estimate be for Russia when k = 16, using Range Normalisation?

In [43]:
# Using the Standardised, copied Data Frame from the previous section 
knn = 16 
estimate = 0 

for i in range(knn):
    estimate += cpi_copy.iloc[i].cpi/(i+1)

estimate = estimate/knn
estimate = estimate * cpi['cpi'].max()

print(estimate)

0.8663388467364912


__Answer: The estimate for Russie's CPI using weighted Nearest Neighbour is 0.8663__

### Russia's CPI
Question 3e. Russia's actual CPI was 2.4488

According to my calculations, using the KNN where K = 3, with a standardised range was most accurate, with a CPI of 3.6585. Using a KNN algorithm where K = 3 resulted in a higher difference, with a CPI of 4.5891, and both versions of the weight KNN algorithm resulted in lower scores with the non standardised estimate being 1.0480 and the standardised estimate falling far too short at 0.8663.

One reason that the weighted scores might have been more sensitive to  outliers such as the US, and the UK, which shared similar feature spaces to Russia but had drastically different CPI values. This was more heavily accentuated when using a standardised scale as this increased those difference by a multitude. Choosing a lower K value helped, however, this was still susceptible to noise, where the top 3 similar values shared very different CPI values. Using a standardised scale helped to smooth out the noise in the data and therefore resulted in a more accurate model. 